### PR TITLE
8349563: Improve AbsNode::Value() for integer types

### PR DESCRIPTION
**Pull request by jaskarth**
Hi all,
This is a small patch that improves the implementation of Value() for `AbsINode` and `AbsLNode` by returning the absolute value of the input range. Previously, C2 would only return an improved type if the input type was a constant. With this change, `Abs` nodes will give a sharper result for all int/long input ranges. The analysis is done in the unsigned domain, and the signed bounds are inferred from the unsigned bounds. As the range of `Abs` is `[0, signed_max+1]` in the unsigned domain, doing the analysis in the unsigned domain allows us to find sharper bounds. I've also added some unit tests and updated the file to limit IR check platforms with more granularity.

Thoughts and reviews would be appreciated!

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[x\] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8349563](https://bugs.openjdk.org/browse/JDK-8349563): Improve AbsNode::Value() for integer types (**Enhancement** - P4)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - **Reviewer**) Review applies to [c9a7a70d](https://git.openjdk.org/jdk/pull/23685/files/c9a7a70da6f549563abed76781f77aaa83066804)
 * [Marc Chevalier](https://openjdk.org/census#mchevalier) (@marc-chevalier - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23685/head:pull/23685` \
`$ git checkout pull/23685`

Update a local copy of the PR: \
`$ git checkout pull/23685` \
`$ git pull https://git.openjdk.org/jdk.git pull/23685/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23685`

View PR using the GUI difftool: \
`$ git pr show -t 23685`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23685.diff">https://git.openjdk.org/jdk/pull/23685.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23685#issuecomment-2667528247)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/subnode.cpp [hunk 1]
```diff
@@ -1928,6 +1928,38 @@ bool BoolNode::is_counted_loop_exit_test() {
...
+static const IntegerType* integral_abs_value(const IntegerType* t) {
+  typedef typename IntegerType::NativeType NativeType;
+
+  assert(!t->is_con(), "Constant types must have already been handled");
+
+  NativeType lo_abs = uabs(t->_lo);
+  NativeType hi_abs = uabs(t->_hi);
+
+  if (lo_abs < 0) {
+    assert(lo_abs == std::numeric_limits<NativeType>::min(), "uabs(t->_lo) must be min value if negative!");
```
Comment (liach) [inline src/hotspot/share/opto/subnode.cpp:1941]: I think asserting `t->_lo` to be min is more straightforward, and also indicates `(t->_lo) + 1`, which yields max, is in the type. We can simplify the comment below too.

File: test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java [hunk 2]
```diff
@@ -229,4 +231,104 @@ public void testChar() {
...
+    public boolean testIntRange2(int in) {
+        // [-4, 3] => [0, 4]
+        return Math.abs((in & 7) - 4) < 0;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.ABS_I })
+    public boolean testIntRange3(int in) {
+        // [-9, -2] => [2, 9]
+        return Math.abs(-((in & 7) + 2)) < 2;
```
Comment (ichttt) [inline test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java:295]: Not sure if this is in scope for this PR, but `abs(x)` should be idealized into `0 - x` if x <= 0. This seems to be missing at the moment.

File: src/hotspot/share/opto/subnode.cpp [hunk 3]
```diff
@@ -1928,6 +1928,38 @@ bool BoolNode::is_counted_loop_exit_test() {
...
 }
 
+template<typename IntegerType>
+static const IntegerType* integral_abs_value(const IntegerType* t) {
+  typedef typename IntegerType::NativeType NativeType;
+
+  assert(!t->is_con(), "Constant types must have already been handled");
+
+  NativeType lo_abs = uabs(t->_lo);
+  NativeType hi_abs = uabs(t->_hi);
```
Comment (dean-long) [inline src/hotspot/share/opto/subnode.cpp:1938]: Converting unsigned to signed is C++ Undefined Behavior, is it not?

Comment (jaskarth) [reply src/hotspot/share/opto/subnode.cpp:1938]: This is a great point, I believe you're correct that it's UB. We currently do the same logic in the old code as well:
https://github.com/openjdk/jdk/blob/a0677d94d8c83a75cee054700e098faa97edca3c/src/hotspot/share/opto/subnode.cpp#L1945-L1947
However, I'm unsure what the best way to solve this would be. Do you happen to have any ideas? Thanks a lot!

File: test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java (hunk 2 above)
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java:295]: This is a good observation! I'll do this in a followup patch, to keep this one focused on just the Value() function.

File: src/hotspot/share/opto/subnode.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/opto/subnode.cpp:1943]: @dean-long  ?

Comment (merykitty) [reply src/hotspot/share/opto/subnode.cpp:1943]: No converting unsigned to signed is not UB, the behaviour is the same as in Java.

Comment (dean-long) [reply src/hotspot/share/opto/subnode.cpp:1943]: I believe it's actually implementation-defined, not UB, until C++ 20, according to discussion in this other PR:
https://github.com/openjdk/jdk/pull/24184#discussion_r2011464234

File: test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java [hunk 4]
```diff
@@ -229,4 +231,104 @@ public void testChar() {
...
+
+        for (int j = 0; j < 20; j++) {
+            int i = Generators.G.ints().next();
+            checkIntRange(i);
+        }
+    }
+
+    @DontCompile
+    public void checkIntRange(int i) {
+        Asserts.assertEquals(Math.abs((i & 7) - 4) > 4, testIntRange1(i));
```
Comment (dafedafe) [inline test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java:249]: Cool improvement @jaskarth!
It might not be directly related to your optimization and marginally relevant but I was wondering if it would make sense to widen the choice of constants a bit (maybe adding few more or some randomly generated ones)?

**Comment by jaskarth** [comment]
Thank you all for the comments! I've pushed an update that refactors the method to check for `min_value` ahead of doing abs, so that we can safely use `ABS()` instead of `uabs()`. I've refactored the behavior for constants to avoid using `uabs` there as well. A re-review would be appreciated!

File: test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java (hunk 4 above)
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java:249]: Thanks for taking a look! Currently we're using the Generators API to create randomly distributed integers to test with (in `checkIntRanges`, from which `checkIntRange` is called) while the constants here are to force the AbsNode to have a range of `[-4, 3]` to ensure that it can be statically optimized away.

**Comment by dafedafe** [comment]
Thanks @jaskarth.
Running tier1-3 tests...

**Review by eme64 — COMMENTED** [review]
@jaskarth Nice work! I have a few comments below.

One is about more randomized tests. I'm thinking about something like this:

- compute `res = Math.abs(x)`
- Truncate `x` with randomly produced bounds from Generators, like this: `x = Math.max(lo, Math.min(hi, x))`.
- Below, add all sorts of comparisons with random constants, like this: `if (res < CON) { sum += 1; }`. If the output range is wrong, this could wrongly constant fold, and allow us to catch that.
- Then fuzz the generated method a few times with random inputs for `x`, and check that the sum and res value are the same for compiled and interpreted code.

I hope that makes sense :)
This is currently my best method to check if ranges are correct, and I think it is quite important because often tests are only written with constants in mind, but less so with ranges, and then we mess up the ranges because it is just too tricky.

This is an example, where I asked someone to try this out as well:
https://github.com/openjdk/jdk/pull/23089/files#diff-12bebea175a260a6ab62c22a3681ccae0c3d9027900d2fdbd8c5e856ae7d1123R404-R422

File: test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java [hunk 5]
```diff
@@ -229,4 +231,104 @@ public void testChar() {
...
+        // [-9, -2] => [2, 9]
+        return Math.abs(-((in & 7) + 2)) < 2;
+    }
+
+    @Test
+    @IR(failOn = { IRNode.ABS_L }, applyIfPlatform = { "64-bit", "true" })
+    public boolean testLongRange4(long in) {
+        // [-9, -2] => [2, 9]
+        return Math.abs(-((in & 7) + 2)) > 9;
+    }
```
[the remark this answers is not in the document]
Comment (jaskarth) [reply test/hotspot/jtreg/compiler/c2/irTests/TestIRAbs.java:558]: I've added support for randomized ranges and if statement folding as suggested in the review comment.

File: src/hotspot/share/opto/subnode.cpp [hunk 6]
```diff
@@ -1933,6 +1933,44 @@ bool BoolNode::is_counted_loop_exit_test() {
...
+  typedef typename IntegerType::NativeType NativeType;
+
+  if (t->is_con()) {
+    // Handle minimum value separately to avoid overflow
+    if (t->get_con() == IntegerType::MIN->_lo) {
+      return IntegerType::MIN;
+    }
+
+    return IntegerType::make(ABS(t->get_con()));
+  }
```
[the remark this answers is not in the document]
Comment (jaskarth) [reply src/hotspot/share/opto/subnode.cpp]: There was an earlier discussion in the review: https://github.com/openjdk/jdk/pull/23685#discussion_r1972735806
Essentially, the implementation of `uabs` relies on converting ints/longs from signed to unsigned which is implementation defined until C++20. I believe the implementation works as expected on most platforms, but to be cautious I thought it would be better to just handle it manually to avoid any potential problems. We should revisit when we're at C++20 ;)

**Comment by jaskarth** [comment]
@eme64 Thanks for the review and comments! The method of checking for constant folding with if statements and range filtering you mentioned is pretty clever. I've adapted it to the test and added it to the PR. Let me know what you think!

**Comment by eme64** [comment]
@jaskarth Nice, thanks for adding the range tests! Unfortunately, I'm quite busy before going on vacation. I hope someone else can review this. Otherwise I can come back to it in August.

**Comment by merykitty** [comment]
With unsigned bounds, you can simply do:

    juint umin = MIN2<juint>(uabs(t->_ulo), uabs(t->_uhi));
    juint umax = MAX2<juint>(uabs(t->_lo), uabs(t->_hi));
    return TypeInt::make_unsigned(umin, umax, t->_widen);

The proof can be inferred trivially from the property of `TypeInt` (you can find this in the doc of `TypeInt`). Since the set of values of a `TypeInt` looks like this:

    smin ---------- lo ===== uhi ------ 0 -----ulo ========= hi --------- smax

or (in this case `lo == ulo`, `hi == uhi`)

    smin ----------- lo ======= hi --- 0 ------------------------------------- smax

or (in this case `lo == ulo`, `hi == uhi`)

    smin --------------------------------- 0 ------- lo ========= hi -------- smax

You can see that in all 3 cases, the minimum of the uabs of a value is either `uabs(ulo)` or `uabs(uhi)` and the maximum is either `uabs(lo)` or `uabs(hi)` (also note that `uabs` cannot overflow, that is the absolute value of a `jint` is always a `juint`)

**Review by merykitty — APPROVED** [review]
Thanks for your change, it looks good to me, I submitted testing.

**Comment by merykitty** [comment]
The tests passed.

File: src/hotspot/share/opto/subnode.cpp [hunk 7]
```diff
@@ -1912,6 +1912,23 @@ bool BoolNode::is_counted_loop_exit_test() {
...
 }
 
+template<typename IntegerType>
+static const IntegerType* integral_abs_value(const IntegerType* t) {
+  typedef typename IntegerType::NativeUType NativeUType;
+
+  // Find the absolute value of a type, resulting in a range that fits inside the unsigned range [0, signed_max+1].
+
+  // An integer type consists of two simple ranges [lo, signed(uhi)] and [signed(ulo), hi] where signed(uhi) <= 0 and
+  // signed(ulo) >= 0. Finding the minimum value of the result type involves finding which of ulo and uhi are closer to 0,
```
Comment (dean-long) [inline src/hotspot/share/opto/subnode.cpp]: That's not quite right.  For example, (int)TypeInt::MAX->_uhi is signed max, which is > 0, and (int)TypeInt::MIN->_ulo is signed min, which is < 0.

Comment (merykitty) [reply src/hotspot/share/opto/subnode.cpp]: Yeah, this only talks about the most sophisticated case and ignores the simple ones. I suggest you bring the diagram I drew above. It is always easier to reason about something if you can visualise it.

File: src/hotspot/share/opto/type.cpp [hunk 8]
```diff
@@ -1799,6 +1799,11 @@ const TypeInt* TypeInt::make(jint lo, jint hi, int widen) {
   return make_or_top(TypeIntPrototype<jint, juint>{{lo, hi}, {0, max_juint}, {0, 0}}, widen)->is_int();
 }
 
+const TypeInt* TypeInt::make_unsigned(juint ulo, juint uhi, int widen) {
+  assert(ulo <= uhi, "must be legal bounds");
+  return make_or_top(TypeIntPrototype<jint, juint>{{min_jint, max_jint}, {ulo, uhi}, {0, 0}}, widen)->is_int();
```
Comment (dean-long) [inline src/hotspot/share/opto/type.cpp:1807]: Is it worth mentioning in a comment that this is overly conservative and relies on canonicalization to clean up?

Comment (jaskarth) [reply src/hotspot/share/opto/type.cpp:1807]: I think it would be worth mentioning to make it more clear, I've pushed a patch that adds a comment. Though for what it's worth, signed `make` does something similar where it gives the unsigned range as `[0, umax]` and the proper unsigned range is inferred from the signed range.

File: src/hotspot/share/opto/subnode.cpp (hunk 7 above)
Comment (jaskarth) [reply src/hotspot/share/opto/subnode.cpp]: Thanks for catching this! I tried to cover the cases concisely, but I think it's too confusing. I've restructured it to show signed and unsigned ranges with the diagrams above, which should help simplify the explanation.

**Comment by jaskarth** [comment]
Thanks for running tests @merykitty, and for taking a look @dean-long! I've pushed a commit that rewords the comments as suggested.

**Review by merykitty — APPROVED** [review]
Thanks for making the changes, it looks good to me.

**Comment by jaskarth** [comment]
May I have some more reviews on this? Thanks a lot!

**Comment by jaskarth** [comment]
May I please have another review on this? Maybe @eme64 or @dean-long? Thanks a lot!

**Comment by marc-chevalier** [comment]
If nobody else does it by then, I'll look at it soon. I've started a bit, but probably won't go much further today.

**Comment by eme64** [comment]
Hi @jaskarth I had another quick look.

--------------------------------

Your PR description says:
> except for the special case where _lo == jint_min/jlong_min which must return the entire type range when encountered

Suppose `t = [min_jint .. -1]`.
```
  _lo  = 0x80000000 = min_jint
  _hi  = 0xffffffff = -1
  _ulo = 0x80000000 = max_jint+1
  _uhi = 0xffffffff = max_juint
umin = min(g_uabs(0x80000000), g_uabs(0xffffffff)) = min(0x80000000, 0x00000001) = 1
umax = max(g_uabs(0x80000000), g_uabs(0xffffffff)) = max(0x80000000, 0x00000001) = 0x80000000
```
And this unsigned range is not a full type range.

-----------------------------------

Do we have tests that deterministically test values close around the "edge-cases"? Such as around 0, `min_jint`, `max_jint`, `max_juint`, etc?

---------------------------------

I'm also wondering if we could keep info about known bits, while doing the optimization? That could also be a follow.-up - if it is even possible.

**Comment by marc-chevalier** [comment]
Yes, the PR description is outdated. Yes, there are some deterministic tests using `ispecial` and `lspecial`.

**Review by marc-chevalier — APPROVED** [review]
It's good. Would just be nice to update the PR description post 19467e3e40ef35248aff7f7a80718fce2f32647f

**Comment by eme64** [comment]
> Yes, there are some deterministic tests using ispecial and lspecial.

@marc-chevalier But do those special values/ranges actually make it to the type `t`? Because it seems to me that the `lspecial` and `ispecial` values are passed into the test methods as `long` or `int`, through a `@DontCompile` method call. So the type info would not actually arrive in `integral_abs_value` for type `t`, right?

**Comment by merykitty** [comment]
@eme64 That description is outdated, you should look only at the description inside the changes.

**Comment by eme64** [comment]
@merykitty Yes, I'm aware that the description is outdated. That's what I'm pointing out, and should be fixed ;)

**Comment by jaskarth** [comment]
Thanks for the review @marc-chevalier, and for taking a look @eme64! I've updated the PR description to better describe the current approach. Does testing need to be run?

**Comment by eme64** [comment]
For me, there is still the question of testing with edge-case type ranges. We have a random test, but some deterministic examples would be great as well. What do you think?

**Comment by marc-chevalier** [comment]
Even if the last run of @merykitty was just on last state up to some comments, it's pretty old. Maybe you can merge master in your branch and I'll run testing.

The new PR description is fine for me. It explains the important part: we go from signed to unsigned, so there is no surprise.

**Comment by jaskarth** [comment]
@marc-chevalier I've updated the branch from the latest master, so it should be good for testing now.
@eme64 I think this is a good point, I've added some more folding tests with constant ranges. Let me know what you think!

**Comment by marc-chevalier** [comment]
Testing is on its way.

**Comment by marc-chevalier** [comment]
Testing looks good.

**Comment by jaskarth** [comment]
Thank you for running testing @marc-chevalier!

**Review by marc-chevalier — APPROVED** [review]
I was happy, I'm still happy.

**Comment by jaskarth** [comment]
Thank you for the re-review @marc-chevalier! I think I need a 2nd review. @eme64 Did the latest commit address your comments about the tests? Thanks!

File: src/hotspot/share/opto/subnode.cpp [hunk 9]
```diff
@@ -1921,6 +1921,31 @@ bool BoolNode::is_counted_loop_exit_test() {
...
+
+  // For these ranges, the result in the unsigned domain is simply [min(|lo|, |hi|), max(|lo|, |hi|)]:
+  // 0-----------min(|lo|,|hi|)==================max(|lo|,|hi|)-----------umax
+
+  NativeUType umin = MIN2<NativeUType>(g_uabs(t->_ulo), g_uabs(t->_uhi));
+  NativeUType umax = MAX2<NativeUType>(g_uabs(t->_lo), g_uabs(t->_hi));
+
+  return IntegerType::make_unsigned(umin, umax, t->_widen);
+}
+
```
Comment (eme64) [inline src/hotspot/share/opto/subnode.cpp:1948]: @jaskarth @merykitty Question: could we have done this using rangeinference.hpp, like in https://github.com/openjdk/jdk/pull/28897? Would that not give us better coverage, if we also add the gtests?

**Review by eme64 — APPROVED** [review]
@jaskarth This looks good now, and thanks to @marc-chevalier for running testing.

I was wondering if we could/should have implemented `integral_abs_value` in `src/hotspot/share/opto/rangeinference.hpp` instead, so we could also perform the more exhaustive `gtest`? @merykitty What do you think? We can also leave that as a follow-up RFE, if @jaskarth does not have more time for it now.

File: src/hotspot/share/opto/subnode.cpp (hunk 9 above)
Comment (merykitty) [reply src/hotspot/share/opto/subnode.cpp:1948]: @eme64 Yes, it could be done. However, I think the inference here is simple enough, and this PR has been in review for a long time, so refactoring it to `RangeInference` could be done as a separate RFE.

Comment (eme64) [reply src/hotspot/share/opto/subnode.cpp:1948]: Sounds good :)

**Comment by jaskarth** [comment]
Thank you all for the reviews!

/integrate
